File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/trending/DefaultTrendingService.java
package dev.reuss.tmdb.domain.trending;

import dev.reuss.tmdb.core.http.QueryParams;
import dev.reuss.tmdb.core.http.TmdbHttpClient;
import dev.reuss.tmdb.core.http.TmdbRequest;
import dev.reuss.tmdb.domain.trending.model.*;
import dev.reuss.tmdb.value.language.Language;

import java.util.Objects;

/**
 * Default {@link TrendingService} implementation backed by the TMDB HTTP client.
 */
public final class DefaultTrendingService implements TrendingService {

    private final TmdbHttpClient httpClient;

    public DefaultTrendingService(TmdbHttpClient httpClient) {
        this.httpClient = Objects.requireNonNull(httpClient, "TMDB HTTP client must not be null");
    }

    @Override
    public TrendingAllResponse all(TrendingTimeWindow timeWindow) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");

        return httpClient.get(
                TmdbRequest.get(TrendingPaths.all(timeWindow)),
                TrendingAllResponse.class
        );
    }

    @Override
    public TrendingAllResponse all(TrendingTimeWindow timeWindow, Language language) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");
        Objects.requireNonNull(language, "Language must not be null");

        return httpClient.get(
                TmdbRequest.get(
                        TrendingPaths.all(timeWindow),
                        QueryParams.create().add("language", language)
                ),
                TrendingAllResponse.class
        );
    }

    @Override
    public TrendingMovieResponse movies(TrendingTimeWindow timeWindow) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");

        return httpClient.get(
                TmdbRequest.get(TrendingPaths.movies(timeWindow)),
                TrendingMovieResponse.class
        );
    }

    @Override
    public TrendingMovieResponse movies(TrendingTimeWindow timeWindow, Language language) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");
        Objects.requireNonNull(language, "Language must not be null");

        return httpClient.get(
                TmdbRequest.get(
                        TrendingPaths.movies(timeWindow),
                        QueryParams.create().add("language", language)
                ),
                TrendingMovieResponse.class
        );
    }

    @Override
    public TrendingPersonResponse people(TrendingTimeWindow timeWindow) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");

        return httpClient.get(
                TmdbRequest.get(TrendingPaths.people(timeWindow)),
                TrendingPersonResponse.class
        );
    }

    @Override
    public TrendingPersonResponse people(TrendingTimeWindow timeWindow, Language language) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");
        Objects.requireNonNull(language, "Language must not be null");

        return httpClient.get(
                TmdbRequest.get(
                        TrendingPaths.people(timeWindow),
                        QueryParams.create().add("language", language)
                ),
                TrendingPersonResponse.class
        );
    }

    @Override
    public TrendingTvShowResponse tv(TrendingTimeWindow timeWindow) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");

        return httpClient.get(
                TmdbRequest.get(TrendingPaths.tvShows(timeWindow)),
                TrendingTvShowResponse.class
        );
    }

    @Override
    public TrendingTvShowResponse tv(TrendingTimeWindow timeWindow, Language language) {
        Objects.requireNonNull(timeWindow, "Trending time window must not be null");
        Objects.requireNonNull(language, "Language must not be null");

        return httpClient.get(
                TmdbRequest.get(
                        TrendingPaths.tvShows(timeWindow),
                        QueryParams.create().add("language", language)
                ),
                TrendingTvShowResponse.class
        );
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/trending/DefaultTrendingService.kt
package dev.reuss.tmdb.domain.trending

import dev.reuss.tmdb.core.http.QueryParams
import dev.reuss.tmdb.core.http.TmdbHttpClient
import dev.reuss.tmdb.core.http.TmdbRequest
import dev.reuss.tmdb.domain.trending.model.*
import dev.reuss.tmdb.value.language.Language

/**
 * Default [TrendingService] implementation backed by the TMDB HTTP client.
 */
internal class DefaultTrendingService(
    private val httpClient: TmdbHttpClient
) : TrendingService {

    override fun all(timeWindow: TrendingTimeWindow): TrendingAllResponse =
        httpClient.get(
            TmdbRequest.get(TrendingPaths.all(timeWindow)),
            TrendingAllResponse::class.java
        )

    override fun all(
        timeWindow: TrendingTimeWindow,
        language: Language
    ): TrendingAllResponse =
        httpClient.get(
            TmdbRequest.get(
                TrendingPaths.all(timeWindow),
                QueryParams.create()
                    .add("language", language.value)
            ),
            TrendingAllResponse::class.java
        )

    override fun movies(timeWindow: TrendingTimeWindow): TrendingMovieResponse =
        httpClient.get(
            TmdbRequest.get(TrendingPaths.movies(timeWindow)),
            TrendingMovieResponse::class.java
        )

    override fun movies(
        timeWindow: TrendingTimeWindow,
        language: Language
    ): TrendingMovieResponse =
        httpClient.get(
            TmdbRequest.get(
                TrendingPaths.movies(timeWindow),
                QueryParams.create()
                    .add("language", language.value)
            ),
            TrendingMovieResponse::class.java
        )

    override fun people(timeWindow: TrendingTimeWindow): TrendingPersonResponse =
        httpClient.get(
            TmdbRequest.get(TrendingPaths.people(timeWindow)),
            TrendingPersonResponse::class.java
        )

    override fun people(
        timeWindow: TrendingTimeWindow,
        language: Language
    ): TrendingPersonResponse =
        httpClient.get(
            TmdbRequest.get(
                TrendingPaths.people(timeWindow),
                QueryParams.create()
                    .add("language", language.value)
            ),
            TrendingPersonResponse::class.java
        )

    override fun tv(timeWindow: TrendingTimeWindow): TrendingTvShowResponse =
        httpClient.get(
            TmdbRequest.get(TrendingPaths.tvShows(timeWindow)),
            TrendingTvShowResponse::class.java
        )

    override fun tv(
        timeWindow: TrendingTimeWindow,
        language: Language
    ): TrendingTvShowResponse =
        httpClient.get(
            TmdbRequest.get(
                TrendingPaths.tvShows(timeWindow),
                QueryParams.create()
                    .add("language", language.value)
            ),
            TrendingTvShowResponse::class.java
        )
}